go/arrow/util/protobuf_reflect.go

```diff
@@ -0,0 +1,499 @@
+	return pfr.descriptor.IsList() && pfr.rValue.Kind() == reflect.Slice
+}
+
+func (pfr protobufFieldReflection) getListLength() int {
+	return pfr.prValue.List().Len()
+}
+
+func (pfr protobufFieldReflection) getMapLength() int {
+	return pfr.prValue.Map().Len()
+}
+
+func (plr protobufListReflection) generateListItems() chan protobufFieldReflection {
+	out := make(chan protobufFieldReflection)
+
+	go func() {
+		defer close(out)
+		for i := 0; i < plr.prValue.List().Len(); i++ {
+			out <- protobufFieldReflection{
+				descriptor:    plr.descriptor,
+				prValue:       plr.prValue.List().Get(i),
+				rValue:        plr.rValue.Index(i),
+				schemaOptions: plr.schemaOptions,
+			}
+		}
+	}()
+
+	return out
+}
+
+func (pfr protobufFieldReflection) getDataType() arrow.DataType {
+	var dt arrow.DataType
+
+	typeMap := map[protoreflect.Kind]arrow.DataType{
+		//Numeric
+		protoreflect.Int32Kind:    arrow.PrimitiveTypes.Int32,
+		protoreflect.Int64Kind:    arrow.PrimitiveTypes.Int64,
+		protoreflect.Sint32Kind:   arrow.PrimitiveTypes.Int32,
+		protoreflect.Sint64Kind:   arrow.PrimitiveTypes.Int64,
+		protoreflect.Uint32Kind:   arrow.PrimitiveTypes.Uint32,
+		protoreflect.Uint64Kind:   arrow.PrimitiveTypes.Uint64,
+		protoreflect.Fixed32Kind:  arrow.PrimitiveTypes.Uint32,
+		protoreflect.Fixed64Kind:  arrow.PrimitiveTypes.Uint64,
+		protoreflect.Sfixed32Kind: arrow.PrimitiveTypes.Int32,
+		protoreflect.Sfixed64Kind: arrow.PrimitiveTypes.Int64,
+		protoreflect.FloatKind:    arrow.PrimitiveTypes.Float32,
+		protoreflect.DoubleKind:   arrow.PrimitiveTypes.Float64,
+		//Binary
+		protoreflect.StringKind: arrow.BinaryTypes.String,
+		protoreflect.BytesKind:  arrow.BinaryTypes.Binary,
+		//Fixed Width
+		protoreflect.BoolKind: arrow.FixedWidthTypes.Boolean,
+		// Special
+		protoreflect.EnumKind:    nil,
+		protoreflect.MessageKind: nil,
+	}
+	dt = typeMap[pfr.descriptor.Kind()]
+
+	switch pfr.arrowType() {
+	case arrow.DICTIONARY:
+		dt = pfr.asDictionary().getDataType()
+	case arrow.LIST:
+		dt = pfr.asList().getDataType()
+	case arrow.MAP:
+		dt = pfr.asMap().getDataType()
+	case arrow.STRUCT:
+		dt = pfr.asStruct().getDataType()
+	}
+
+	return dt
+}
+
+func RecordFromProtobuf(psr ProtobufStructReflection, schema *arrow.Schema, mem memory.Allocator) arrow.Record {
+	if mem == nil {
+		mem = memory.NewGoAllocator()
+	}
+
+	recordBuilder := array.NewRecordBuilder(mem, schema)
+
+	var fieldNames []string
+	for i, field := range schema.Fields() {
+		AppendValueOrNull(recordBuilder.Field(i), psr.getFieldByName(field.Name), field, mem)
+		fieldNames = append(fieldNames, field.Name)
+	}
+
+	var arrays []arrow.Array
+	for _, bldr := range recordBuilder.Fields() {
+		a := bldr.NewArray()
+		arrays = append(arrays, a)
+	}
+
+	structArray, _ := array.NewStructArray(arrays, fieldNames)
+
+	return array.RecordFromStructArray(structArray, schema)
+}
+
+func AppendValueOrNull(b array.Builder, pfr protobufFieldReflection, f arrow.Field, mem memory.Allocator) {
```

Comment:
should this really be externally exported as opposed to kept as an internal only function?